File: src/org/jgll/sppf/SPPFNode.java
package org.jgll.sppf;

import java.util.Iterator;

import org.jgll.traversal.SPPFVisitor;

/**
 * An SPPF node is a node in an Shared Packed Parse Forest. This data structure
 * is used to store the parse forest that is the result of parsing an input
 * string with a GLL parser. <br />
 * 
 * @author Maarten Manders
 * @author Ali Afroozeh
 * 
 */

public abstract class SPPFNode implements Iterable<SPPFNode> {

	private boolean visited;

	private Object object;

	public abstract String getId();

	public abstract String getLabel();

	public abstract SPPFNode get(int index);

	public abstract int size();

	public abstract int getLeftExtent();

	public abstract int getRightExtent();

	public abstract void accept(SPPFVisitor visitAction);

	public Iterable<Object> childrenValues() {

		final Iterator<SPPFNode> iterator = iterator();

		return new Iterable<Object>() {

			@Override
			public Iterator<Object> iterator() {
				return new Iterator<Object>() {

					private SPPFNode next;

					@Override
					public boolean hasNext() {
						while (iterator.hasNext() && (next = iterator.next()).getObject() != null) {
							return true;
						}
						return false;
					}

					@Override
					public Object next() {
						return next.getObject();
					}

					@Override
					public void remove() {
						throw new UnsupportedOperationException();
					}
				};

			}
		};
	}

	public boolean isVisited() {
		return visited;
	}

	public void setVisited(boolean visited) {
		this.visited = visited;
	}

	public Object getObject() {
		return object;
	}

	public void setObject(Object object) {
		this.object = object;
	}
}

File: src-scala/org/jgll_staged/sppf/SPPFNode.scala
package org.jgll_staged.sppf

import java.util.Iterator
import org.jgll_staged.grammar.slot.GrammarSlot
import org.jgll_staged.traversal.SPPFVisitor
import org.jgll_staged.util.hashing.Level
import scala.reflect.{BeanProperty, BooleanBeanProperty}
//remove if not needed
import scala.collection.JavaConversions._

abstract class SPPFNode extends Level {

  @BooleanBeanProperty
  var visited: Boolean = _

  @BeanProperty
  var `object`: AnyRef = _

  def getLabel(): String

  def getChildAt(index: Int): SPPFNode

  def getChildren(): java.lang.Iterable[SPPFNode]

  def isAmbiguous(): Boolean

  def childrenCount(): Int

  def getLeftExtent(): Int

  def getRightExtent(): Int

  def accept(visitAction: SPPFVisitor): Unit

  def getGrammarSlot(): GrammarSlot

  def deepEquals(node: SPPFNode): Boolean = {
    if (this != node) {
      return false
    }
    if (this.childrenCount() != node.childrenCount()) {
      return false
    }
    if (this.isAmbiguous ^ node.isAmbiguous) {
      return false
    }
    if (this.isAmbiguous && node.isAmbiguous) {
      val thisIt = getChildren.iterator()
      outer: while (thisIt.hasNext) {
        val thisChild = thisIt.next()
        val otherIt = node.getChildren.iterator()
        while (otherIt.hasNext) {
          val otherChild = otherIt.next()
          if (thisChild.deepEquals(otherChild)) {
            //continue
          }
        }
        return false
      }
      return true
    }
    val thisIt = getChildren.iterator()
    val otherIt = node.getChildren.iterator()
    while (thisIt.hasNext && otherIt.hasNext) {
      val thisChild = thisIt.next()
      val otherChild = otherIt.next()
      if (!thisChild.deepEquals(otherChild)) {
        return false
      }
    }
    true
  }
}
